### PR TITLE
Use mongoose to validate method to check for ID validity

### PR DESCRIPTION
**Pull request by carlbennettnz**
Which is better than assuming every _id field needs an ObjectId.

This method is slightly inefficient as it does try to validate every other field in the schema as well. If there's a way to limit mongoose to validating a single field, we should do that, but this is the best method I've found so far.

Potential solution to #80

### CODE REVIEW

**Comment by ethanresnick** [comment]
Thanks for submitting this. I'll try to take a look at it soon but, as I'm sure you've noticed, I haven't had much time for this project recently. Is there any chance you can reformat these changes to make the diff as small as possible? That will make it much easier for me to follow what's changed and get things merged quickly.

**Comment by carlbennettnz** [comment]
> Is there any chance you can reformat these changes to make the diff as small as possible?

It's not actually as bad as it looks. I had to change `getIdQueryType` to return a promise, so now half the file is inside `.then()` calls and therefore indented further. Pull the branch and run `git diff --ignore-space-change HEAD^1 HEAD` and you'll get a much nicer diff.

Not sure why that one test is failing. I've spent so much time trying to figure out this project's tests and CI system, but it still does stuff I don't understand. They pass for me locally, even with the same node and npm versions. The only other thing I can think of to reproduce the failure is using Ubuntu rather than OS X, but I have no idea why that would be an issue.

**Comment by carlbennettnz** [comment]
TIL you can add `w=1` to GitHub's URL params to see the diff with whitespace ignored

https://github.com/ethanresnick/json-api/pull/113/files?diff=unified&w=1

**Comment by ethanresnick** [comment]
Thanks Carl!

The smaller diff definitely helps. I'll try to add comments inline. Also, can you add some new tests for the case that this change is supposed to address?

File: src/db-adapters/Mongoose/MongooseAdapter.js [hunk 1]
```diff
@@ -603,27 +604,41 @@ export default class MongooseAdapter {
...
+      const mode = typeof idOrIds === "string" ? "findOne" : "find";
+      let ids = Array.isArray(idOrIds) ? idOrIds : [ idOrIds ];
+      let idQuery;
+      let tests = [];
 
-    if (typeof idOrIds === "string") {
-      if (!this.idIsValid(idOrIds)) {
-        throw new APIError(404, undefined, "No matching resource found.", "Invalid ID.");
-      }
+      ids = ids.filter(id => Boolean(id));
```
Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: What is this line for?

Comment (carlbennettnz) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Remove falsely values. because f we passed `null` as the `_id` when validating then the validation passes. Could maybe be changed to

     ids = ids.filter(id => typeof id === "string");

or we could change the change the line 617 to

    const testDoc = new model({ _id: String(id) });

Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Both of those seem like they'd reject the integer 0 as an id. But 0 should be allowed right, if the point is to go beyond ObjectIds? I think it's legal in mongo.

Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: The perfect is the enemy of the good, though, so I could also be persuaded to leave this as is (i.e. with just the falsey check), so long as we documented that. (I.e., "Ids can be ObjectIds or any truthy value".)

File: src/db-adapters/Mongoose/MongooseAdapter.js [hunk 2]
```diff
@@ -603,27 +604,41 @@ export default class MongooseAdapter {
...
+      ids.forEach(id => {
+        const testDoc = new model({ _id: id })
+        tests.push(testDoc.validate());
+      });
 
-    else if (Array.isArray(idOrIds)) {
-      if (!idOrIds.every(this.idIsValid)) {
-        throw new APIError(400, undefined, "Invalid ID.");
-      }
+      Q.allSettled(tests).then(results => {
```
Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: We shouldn't have to wait for all the validations to complete before determining whether `invalid` is true. Maybe have an outer promise representing the `invalid` flag, and have the first id test that resolves to false reject that outer test.

File: src/db-adapters/Mongoose/MongooseAdapter.js [hunk 3]
```diff
@@ -603,27 +604,41 @@ export default class MongooseAdapter {
...
-    if (typeof idOrIds === "string") {
-      if (!this.idIsValid(idOrIds)) {
-        throw new APIError(404, undefined, "No matching resource found.", "Invalid ID.");
-      }
+      ids = ids.filter(id => Boolean(id));
 
-      idQuery = {_id: idOrIds};
-    }
+      ids.forEach(id => {
+        const testDoc = new model({ _id: id })
```
Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Can we move this logic and the logic on line 625 about checking the reason into a single function, so it's clearer how they connect?

In particular, I'm thinking maybe the existing `idIsValid` function could be changed to:

```
idIsValid(id, model) { 
    return new model({_id: id}).validate().then((result) => !result.reason || !result.reason.errors._id)
}
```

File: src/db-adapters/Mongoose/MongooseAdapter.js (hunk 2 above)
Comment (carlbennettnz) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: If there are any other fields in the model which are required, the validation promise will reject, even if the ID is valid. In this case, resolution/rejection doesn't mean anything useful.

File: src/db-adapters/Mongoose/MongooseAdapter.js (hunk 1 above)
Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Edit: actually, we probably shouldn't just silently filter out invalid ids; instead, they should trigger an error too, right? In which case this check could also be incorporated into the definition of `idIsValid`

Comment (carlbennettnz) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: For `idIsValid`, yes, `null` should reject. It's a bit more complicated for `getIdQueryType` though because a `null` input is used to say no ID query should be returned, and therefore all records will eventually be resolved.

File: src/db-adapters/Mongoose/MongooseAdapter.js (hunk 2 above)
Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Right. By an id test resolving to false, I meant not just that validation fails, but also that the reason for failure was an error with the id. Connecting that with my definition of isIdValid from [here](https://github.com/ethanresnick/json-api/pull/113#discussion_r55955256), it might look something like:

```js
// assuming idIsValid builds in the full validation check and the truthy filter.
const idValidationPromises = Q.all(ids.map((id) => this.idIsValid(id, model));

// the `then` code here will run as soon immediately once/if an id is reported invalid,
// rather than waiting for every validate check to finish like `allSettled` would have to.
return idValidationPromises.then(() => { 
  /* return the query */ 
}, () => 
  { /* throw a 4xx */ 
});
```

Comment (carlbennettnz) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: I've just pushed some new commits which I think do what you're talking about here

**Comment by carlbennettnz** [comment]
Somehow that test that used to fail of Circle has starting passing. It's just eslint that's failing now.

File: src/db-adapters/Mongoose/MongooseAdapter.js (hunk 1 above)
Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Good call. But I think that case only applies if a single id is passed in. (I.e., if `idOrIds` is `[null]`, that should be an error, right?) So I think I'd replace the filter with a conditional that only checks when a single ids is passed in. (I'm not sure what the resolution there should be though; see comment.) 

Also, it'd be nice to clarify that `mode` and `ids` are based on the same isArray check. Finally, I think `idIsValid` should resolve to true, not to the id, when the id is valid (since it has "is" in the name). 

Putting those together:

```
 static getIdQueryType(idOrIds) {
    const [mode, ids] =  Array.isArray(idOrIds) ? ["find", idOrIds] : ["findOne", [idOrIds]];
    
    if(mode === "findOne" && idOrIds == null) {
      // what should this resolution be? 
      // ["findOne", {_id: undefined}] or ["findOne", undefined]
      // I think it was the form before this pr, but is now the latter
    }
    
    const idValidationPromises = ids.map(id => this.idIsValid(id, model));

   return Q.all(idValidationPromises).then(() => {
     const idQuery = { _id: mode === "find" ? { $in: ids } : ids[0] };
     return [ mode, idQuery];
   }, (err) => {
     // ....
   });
});

Comment (carlbennettnz) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Okay, that method is nicer. Pushed.

I'm not sure which resolution makes more sense, but the latter makes the tests pass.

Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Thanks Carl! Re the resolution, maybe resolving with `["find", undefined]` is correct? (i.e. not using findOne at all). Does that make the tests pass?

(I updated code the in comment above accordingly.)

Comment (carlbennettnz) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Pushed

**Comment by carlbennettnz** [comment]
Linting now passes, but coverall is failing. I've never used coverall, so I haven't got a clue what the problem is. This might be helpful: lemurheavy/coveralls-public#632

File: src/db-adapters/Mongoose/MongooseAdapter.js [hunk 4]
```diff
@@ -603,30 +604,40 @@ export default class MongooseAdapter {
...
-        throw new APIError(400, undefined, "Invalid ID.");
-      }
-
-      idQuery = {_id: {"$in": idOrIds}};
-    }
+    const [ mode, ids ] = Array.isArray(idOrIds) ? [ "find", idOrIds ] : [ "findOne", [ idOrIds ] ];
+    const idValidationPromises = ids.map(id => this.idIsValid(id, Model));
 
-    return [mode, idQuery];
+    return Q.all(idValidationPromises).then((idsAreValid) => {
```
Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Ugh, now we're back to waiting for all the validations to finish before we run the `then` block, even if we could skip to throwing an error earlier. 

This is my fault... I said earlier that, if we're going to keep `isIdValid` named as it is, it should resolve with true or false. But, what I meant was that it should *resolve* with true if the id is valid but *reject* with false otherwise. That way, the basic structure of `getIdQueryType` can be as it was in my last comment.

I realize that rejecting with false is a bit weird semantically, though, which is why I edited my earlier comment to say that it's probably better to just rename `isIdValid`. So that's what I'd do now.

I.e., I'd rename `isIdValid` to `validateId`, which can then resolve/reject as you had originally (i.e., with the valid id on success, and the validation error on failure). And then the `getIdQueryType` can work with the two different handlers.

**Comment by ethanresnick** [comment]
@carlbennettnz Don't worry about coveralls. It's going to fail for you no matter what you do, because of a permissions error. (Specifically, my coveralls API key isn't available to your code when CircleCI runs it.) This makes some sense, but it's annoying. I opened an issue somewhere to see if this could be worked around, but there hasn't been any movement.

File: src/db-adapters/Mongoose/MongooseAdapter.js (hunk 4 above)
Comment (carlbennettnz) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Ahhh, right. I think I finally understand what you were saying earlier. Yes, we should be failing as soon as possible. We could maintain the semantics of `idIsValid()` by doing something like this:

```js
    const idValidationPromises = ids.map(id => this.idIsValid(id, Model).then(isValid => {
      if (!isValid) {
        throw new Error('ID is invalid');
      }
    }));
```

Do you think retaining the semantics of `idIsValid` is worth the extra complexity in `getIdQueryType`?

Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: > We could maintain the semantics of idIsValid() by doing something like this

That's a pretty clever solution! I like it.

Still, though, I think I'd slightly prefer just renaming `idIsValid` so that settling the promise with richer values (i.e. the id and the error, respectively) then makes sense. I lean that way only because we have the validation error information within that function, so we might as well not throw it away (in case future callers want to do something with it).

Comment (carlbennettnz) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: :+1: 

Pushed

**Comment by ethanresnick** [comment]
Awesome! Thanks @carlbennettnz! I'm heading into a meeting now, but I'll take a final look at this later tonight. I think it should be good to go!

**Comment by carlbennettnz** [comment]
Sweet. Do want me to merge commits now, or after you get a chance to review?

File: src/db-adapters/Mongoose/MongooseAdapter.js [hunk 5]
```diff
@@ -603,30 +604,37 @@ export default class MongooseAdapter {
...
+    }
+
+    const [ mode, ids ] = Array.isArray(idOrIds) ? [ "find", idOrIds ] : [ "findOne", [ idOrIds ] ];
+    const idValidationPromises = ids.map(id => this.validateId(id, Model));
 
-    if (typeof idOrIds === "string") {
-      if (!this.idIsValid(idOrIds)) {
+    return Q.all(idValidationPromises).then(() => {
+      const idQuery = { _id: mode === "find" ? { $in: ids } : ids[0] };
+      return [ mode, ids.length ? idQuery : undefined ];
```
Comment (ethanresnick) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Are you sure this `ids.length ? idQuery : undefined` ternary is needed? 

If I read correctly, it means that calling `getIdQueryOrType` with `[]` will resolve to `["find", undefined]`, which I don't think is what we want, since I'd interpret an array of ids—even if it's empty—to be about returning docs whose ids are from that array, rather than interpreting it as a findAll query. (This seems pedantic, but I can imagine cases where it might come up.)

File: test/app/database/models/numeric-id.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import mongoose from "mongoose";
+
+const schema = mongoose.Schema({
+  _id: {
+    type: Number,
+    required: true
```
Comment (ethanresnick) [inline test/app/database/models/numeric-id.js]: Just out of curiosity, if `_id` isn't required here, does it effect the tests?

Also, in JSON API, a resource's id needs to be a string....so I'm not sure how we want to handle that. For now, it might make sense to only allow strings or ObjectIds.

File: test/app/database/models/string-id.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import mongoose from "mongoose";
+
+const schema = mongoose.Schema({
+  _id: {
+    type: String,
+    required: true
```
Comment (ethanresnick) [inline test/app/database/models/string-id.js]: Just out of curiosity, if `_id` isn't required here, does it effect the tests?

File: test/unit/db-adapters/Mongoose/MongooseAdapter.js [hunk 8]
```diff
@@ -69,68 +74,190 @@ describe("Mongoose Adapter", () => {
...
+          MongooseAdapter.validateId(null, School),
+          MongooseAdapter.validateId(null, NumericId),
+          MongooseAdapter.validateId(null, StringId),
+          MongooseAdapter.validateId(undefined, School),
+          MongooseAdapter.validateId(undefined, NumericId),
+          MongooseAdapter.validateId(undefined, StringId)
+        ];
+
+        Q.allSettled(tests).then((res) => {
+          res.forEach(result => expect(result.state).to.equal("rejected"));
```
Comment (ethanresnick) [inline test/unit/db-adapters/Mongoose/MongooseAdapter.js]: I'm pretty sure es6 promises don't have a `.state` property, and I'd like to be able to switch out q for the real ES6 promises soon. So, is there a way to rewrite this test without checking `.state`?

File: test/unit/db-adapters/Mongoose/MongooseAdapter.js [hunk 9]
```diff
@@ -69,68 +74,190 @@ describe("Mongoose Adapter", () => {
...
+          MongooseAdapter.validateId("one", NumericId),
+          MongooseAdapter.validateId([], NumericId),
+          MongooseAdapter.validateId(true, NumericId),
+          MongooseAdapter.validateId(false, NumericId)
+          // StringId should except anything != null
+        ];
+
+        Q.allSettled(tests).then((res) => {
+          res.forEach(result => {
+            expect(result.state).to.equal("rejected");
```
Comment (ethanresnick) [inline test/unit/db-adapters/Mongoose/MongooseAdapter.js]: same as above

File: test/unit/db-adapters/Mongoose/MongooseAdapter.js [hunk 10]
```diff
@@ -69,68 +74,190 @@ describe("Mongoose Adapter", () => {
...
+      it("should refuse an empty string", function(done) {
+        const tests = [
+          MongooseAdapter.validateId("", School),
+          MongooseAdapter.validateId("", NumericId),
+          MongooseAdapter.validateId("", StringId)
+        ];
+
+        Q.allSettled(tests).then((res) => {
+          res.forEach(result => {
+            expect(result.state).to.equal("rejected");
```
Comment (ethanresnick) [inline test/unit/db-adapters/Mongoose/MongooseAdapter.js]: same as above

File: test/app/database/models/numeric-id.js (hunk 6 above)
Comment (ethanresnick) [inline test/app/database/models/numeric-id.js]: Or maybe we just need to add something in the serialization layer to make sure that, even if the id is a number, it's always serialized as a string? (including in linkage)

File: src/db-adapters/Mongoose/MongooseAdapter.js (hunk 5 above)
Comment (carlbennettnz) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: You're right. It's left over from a previous version

**Comment by ethanresnick** [comment]
Just finished making notes! Re merging the commits now or later: I don't mind either way, but it probably makes sense to wait until all the code's finalized, just so you don't have to do it again if we catch any last bits to change

File: test/app/database/models/numeric-id.js (hunk 6 above)
Comment (carlbennettnz) [inline test/app/database/models/numeric-id.js]: I would go with the second option. I can imagine use cases where someone has an existing schema with numeric IDs and they want to put a JSON API-style API in front of it.

File: test/app/database/models/string-id.js (hunk 7 above)
Comment (carlbennettnz) [inline test/app/database/models/string-id.js]: Yes, the empty string test

File: test/app/database/models/numeric-id.js (hunk 6 above)
Comment (ethanresnick) [inline test/app/database/models/numeric-id.js]: Sounds good to me!

If we do that, though, I think we should make sure the serialization's working before/as part of merging in support for Number ids. So, do you want to add that serialization stuff to this PR? Or keep this PR just to strings and ObjectIds, and then make a second that deals with Numbers (validation/queries + serialization)? I don't mind either way, but I could understand if you're feeling anxious to get this merged and so would prefer the former.

File: test/unit/db-adapters/Mongoose/MongooseAdapter.js (hunk 8 above)
Comment (carlbennettnz) [inline test/unit/db-adapters/Mongoose/MongooseAdapter.js]: Not really. ES6 Promises don't include an `.allSettled` method, so the only options I can think of in this case as reimplementing `.allSettled` in place or continuing to use a promises library for stuff like this along with proper ES6 promises when the library isn't needed. Personally, I've been using RSVP alongside ES6 and it's worked great for me.

File: src/db-adapters/Mongoose/MongooseAdapter.js (hunk 5 above)
Comment (carlbennettnz) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Wait, wouldn't the proper test be `idOrIds != null ? idQuery : undefined`? If you pass `[]` you get an empty query, if you pass `null` you get `undefined`.

Comment (carlbennettnz) [inline src/db-adapters/Mongoose/MongooseAdapter.js]: Pretty sure that's right. Pushed.

File: test/unit/db-adapters/Mongoose/MongooseAdapter.js (hunk 8 above)
Comment (ethanresnick) [inline test/unit/db-adapters/Mongoose/MongooseAdapter.js]: What about something like this?

```
const tests = [/*... */];
const allTestsRejectedPromise = Q.Promise(function(resolve, reject) {
  let rejectedCount = 0;
  const incrementRejectedCount = () => { 
    rejectedCount++; 
    if(rejectedCount === tests.length) {
      resolve();
    }
  }
  
  // if any test promise *resolves*, we *reject* the outer promise,
  // which represents whether all the tests were *rejected*. 
  // And, if the test rejects as expected, we increment a counter which 
  // will resolve the outer promise once all the tests are rejected.
  tests.forEach((testPromise) => testPromise.then(reject, incrementRejectedCount));
}
```

**Comment by ethanresnick** [comment]
P.S. I'm heading to bed now, but will take another look at this in the morning :)

File: test/unit/db-adapters/Mongoose/MongooseAdapter.js (hunk 8 above)
Comment (carlbennettnz) [inline test/unit/db-adapters/Mongoose/MongooseAdapter.js]: Well yes, it's just that's a very common pattern, so common that we have third party libraries that do that and lots of other useful things. I'm not sure what value you get out of reimplementing.

File: test/app/database/models/numeric-id.js (hunk 6 above)
Comment (carlbennettnz) [inline test/app/database/models/numeric-id.js]: Wouldn't `serializedId = String(unserializedId)` cover 99% of use cases? If you're thinking of providing a full serialization API, that could be useful, but surely that's a tiny, tiny edge case.

File: test/unit/db-adapters/Mongoose/MongooseAdapter.js (hunk 8 above)
Comment (ethanresnick) [inline test/unit/db-adapters/Mongoose/MongooseAdapter.js]: > I'm not sure what value you get out of reimplementing.

Having to write just a few lines in order to remove a larger library is worth it imo; it means one less API for me to remember, and for contributors to learn. But there's also a conceptual issue here: while `allSettled` is useful, and my code is in some sense reimplementing that, it's also removing the need for `.state`, which is my main concern. The state of a promise is, by design, not meant to be synchronously inspectable, so q exposing that, and us relying on it, seems like bad practice.

File: test/app/database/models/numeric-id.js (hunk 6 above)
Comment (ethanresnick) [inline test/app/database/models/numeric-id.js]: > Wouldn't serializedId = String(unserializedId) cover 99% of use cases?

Yup, and that's basically what I'm thinking. That logic just has to be applied (at least in the Document type, but maybe other places I'm not thinking of).
